README.md
![alt XOOPS CMS](https://xoops.org/images/logoXoops4GithubRepository.png)
## Soapbox module for  [XOOPS CMS 2.5.9+](https://xoops.org)
[![XOOPS CMS Module](https://img.shields.io/badge/XOOPS%20CMS-Module-blue.svg)](https://xoops.org)
[![Software License](https://img.shields.io/badge/license-GPL-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-2.0.html)

[![Scrutinizer Code Quality](https://img.shields.io/scrutinizer/g/XoopsModules25x/soapbox.svg?style=flat)](https://scrutinizer-ci.com/g/XoopsModules25x/soapbox/?branch=master)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/95b12220e0ac4056b9af52af708379c9)](https://www.codacy.com/app/mambax7/soapbox_2)
[![Code Climate](https://img.shields.io/codeclimate/github/XoopsModules25x/soapbox.svg?style=flat)](https://codeclimate.com/github/XoopsModules25x/soapbox)
[![SensioLabsInsight](https://insight.sensiolabs.com/projects/68207475-07ff-4567-a282-6e2f119077d2/mini.png)](https://insight.sensiolabs.com/projects/68207475-07ff-4567-a282-6e2f119077d2)
[![Latest Pre-Release](https://img.shields.io/github/tag/XoopsModules25x/soapbox.svg?style=flat)](https://github.com/XoopsModules25x/soapbox/tags/)
[![Latest Version](https://img.shields.io/github/release/XoopsModules25x/soapbox.svg?style=flat)](https://github.com/XoopsModules25x/soapbox/releases/)

Soapbox module for [XOOPS CMS](https://xoops.org) that has a single purpose: to help you build in your site an editorial column section, where you can manage the columns and articles published in an ordered fashion.

[![Tutorial Available](https://xoops.org/images/tutorial-available-blue.svg)](https://xoops.gitbook.io/soapbox-tutorial/) Tutorial: see [GitBook](https://xoops.gitbook.io/soapbox-tutorial/).
To contribute to the Tutorial, [fork it on GitHub](https://github.com/XoopsDocs/soapbox-tutorial)

[![Translations on Transifex](https://xoops.org/images/translations-transifex-blue.svg)](https://www.transifex.com/xoops)

Please visit us on https://xoops.org

Current and upcoming "next generation" versions of XOOPS CMS are being crafted on GitHub at: https://github.com/XOOPS


docs/changelog.txt
<h5>1.70 Beta 2      [NOT RELEASED]</h5>  Dev: Xoops 2.5.10, PHP 7.3.2
<hr>
- clear sbvotedata table when adding test data (mamba)
- namespaces/autoloading refactoring (mamba)
- fixes (aerograf)
- added feedback (mamba)
- added test data button (mamba)
- added migrate (mamba)
- added test data (mamba)

<h5>1.70 Beta 1      [2017-05-23]</h5>
<hr>
- XOOPS 2.5.9, PHP 7 (mamba)
- removed @version (mamba)
- PSR-2 code cosmetics (mamba)
- addNavigation(basename(__FILE__))  (mamba)
- dirname(__FILE__) to __DIR__ (mamba)
- replaced < br /> with < br > (mamba)
- Unnecessary double quotes (mamba)
- reference mismatch (mamba)
- Hardening to type safe (mamba)
- update Help (mamba)
- changed _handler to Handler (mamba)
- removed '/' from void HTML elements (&lt;br>, &lt;img>, etc) (mamba)
- HTML 5: removed 'selected' and 'checked' (mamba)
- updated file headers for consistency (mamba)
- renamed several $indexAdmin with $adminObject for consistency (mamba)
- Updates for PayPal changes (zyspec)
- fixed rating check (aerograf)
- moved functions to Utility class (mamba)
- started conversion to XMF (mamba)
    - updated addInfoBoxLine entries (mamba)
    - displayNavigation(basename(__FILE__))  (mamba)
    - changed to ->displayButton('left') (mamba)
- CSS file location fixes (aerograf)

<h5>1.61 Beta 1      [NOT RELEASED]</h5>
<hr>
- moved all images, CSS, and JS files to /assets (mamba)
- renamed .html Smarty templates to .tpl (mamba)
- changed intval() to (int) (mamba)
- changed dirname(__FILE__) to __DIR__ (mamba)
- code cosmetics (mamba)
- naming consistently language variables (mamba)
- Fixing Rating (trabis)


<h5>1.60            [2012-12-18]</h5>
<hr>
- Conversion to XOOPS 2.5.0 GUI (mamba)
- converting to dynamic Editor selection (mamba)
- fixing date issues-


<h5>1.50 23     [August 2004]</h5>
<hr>
This is version 1.5 of Soapbox. Actually, this is NOT the full version, but a stripped down one. It includes all the necessary files to run the module, but for the sake of keeping the size small, it has left out:

+ The language files, which will be available in a separate download once the debugging is complete.
+ The help files, which I still have to update and which will also be available in a separate download in different formats.
For a few days, this will be available as a release candidate, while xoopsers look for bugs and I work on the docs. If you find bugs, please report them to hsalazar@xoops.org. I'll try to fix them if possible.

I've evaluated the inclusion of some other functions, as well as the possible implementation of several suggestions. Her's a couple of comments about them:

+ Most suggestions indicate that xoopsers want to use Soapbox as a general article management module, which is not its intended use. That's why they keep asking for subcategories, several authors per column, expiration dates and such. Soapbox was designed to be used as an editorial column module, with columns being defined as single-author contributions to a site.
+ I've tried to include an internal search function, but decided against it for two reasons: first, XOOPS internal function is good enough for most purposes; second, haven't found a way to paginate results when there are many.
+ I built a version including SPAW as WYSIWYG editor, but decided against it because of the size. Then, I learned some of XOOPS' brightest are working on a cross-browser WYSIWYG editor and decided to wait for their implementation before including such a feature in Soapbox.

Changes for this release:

+ Tested more carefully the blocks, to make sure they don't cause more problems. It might still have surprises, though.
+ Optimized the code updating references to server variables and other stuff.
+ Used Mithrandir's code to optimize the menu in the admin side.
+ Included my own version of Predator's addition: a breadcrumb for the admin side, with a few style tweaks. For next release, will copy this idea to implement it in the user side.
+ Included a function to eliminate image tags from blocks, so now you'll only have images in the articles themselves, not in the teasers, blocks or any other calls.
+ Modified the constant strings so that the module configuration options are more easily read.
+ Included several suggestions by domifara.
+ Corrected the variables needed to be able to change the module's name in the admin section.
+ Added another block: Soapbox spotlight.
+ Included new versions of the admin icons.
+ Modified the admin section to invoke tables using functions.
+ In the admin side, the columns page now includes the list of columns.
+ In the admin side, the articles page now includes the list of articles.
+ In the admin side, changed a bit the behavior of the submissions page.


<h5>1.40 17     [March 2004]</h5>
<hr>
+ Fixed problems that produced blank page on block activation.


<h5>1.20 26     [February 2004]</h5>
<hr>
More bugfixes and small changes

16 March 2004 v1.30
+ Supposedly I repaired the blocks stuff. The idea is this: you, as user, only see those blocks configured in blocks admin for you to see. And what do you see in those blocks? Well, only those articles belonging to columns you're authorized to see. I've checked and rechecked and in my installation it works. Does it work for you?
+ Rebuilt the module so that now the styles are in an external file and are thus easier to change. This is definitely not streamlined, but should make things a bit easier should you want to change the look of the module.
+ Added more options in the user side: icons in the index and column files to "print" and "send to friend". The admin sees also two more icons to "edit" and "delete".
+ Fixed a problem with the autoteaser.
+ In the admin side, added files to manage the module's blocks without leaving the module's environment. These files were contributed (unknowingly) by G. I. Joe.
+ Please check if it still crashes the browser often. It doesn't happen in my installation, but it might in yours...

article.php
<?php
/**
 * Module: Soapbox
 * Version: v 1.5
 * Release Date: 23 August 2004
 * Author: hsalazar
 * Licence: GNU
 */

use Xmf\Request;
use XoopsModules\Soapbox;

require_once __DIR__ . '/header.php';

/** @var Soapbox\Helper $helper */
$helper = Soapbox\Helper::getInstance();

$cleantags = new Soapbox\Cleantags();

$xoopsConfig['module_cache']             = 0; //disable caching since the URL will be the same, but content different from one user to another
$GLOBALS['xoopsOption']['template_main'] = 'sb_article.tpl';
require_once XOOPS_ROOT_PATH . '/header.php';
global $xoopsModule;
//$pathIcon16 = $xoopsModule->getInfo('sysicons16');
$pathIcon16 = \Xmf\Module\Admin::iconUrl('', 16);

$moduleDirName = $myts->htmlSpecialChars(basename(__DIR__));
if ('soapbox' !== $moduleDirName && '' !== $moduleDirName && !preg_match('/^(\D+)(\d*)$/', $moduleDirName)) {
    echo('invalid dirname: ' . htmlspecialchars($moduleDirName, ENT_QUOTES));
}
//---GET view sort --
$sortname = isset($_GET['sortname']) ? mb_strtolower(trim(strip_tags($myts->stripSlashesGPC($_GET['sortname'])))) : 'datesub';
if (!in_array($sortname, ['datesub', 'weight', 'counter', 'rating', 'headline'], true)) {
    $sortname = 'datesub';
}
$sortorder = isset($_GET['sortorder']) ? mb_strtoupper(trim(strip_tags($myts->stripSlashesGPC($_GET['sortorder'])))) : 'DESC';
if (!in_array($sortorder, ['ASC', 'DESC'], true)) {
    $sortorder = 'DESC';
}
//---------------
//require_once XOOPS_ROOT_PATH . '/modules/' . $moduleDirName . '/include/cleantags.php';
//for ratefile update by domifara
require_once XOOPS_ROOT_PATH . '/class/xoopsformloader.php';
//require_once XOOPS_ROOT_PATH . '/modules/' . $moduleDirName . '/include/gtickets.php';

$articleID = Request::getInt('articleID', 0, 'GET'); //isset($_GET['articleID']) ? (int)($_GET['articleID']) : 0;
$startpage = Request::getInt('page', 0, 'GET'); //isset($_GET['page']) ? (int)($_GET['page']) : 0;
//-------------------------------------
//move here  form ratefile.php
if (\Xmf\Request::hasVar('submit', 'POST') && !empty($_POST['lid'])) {
    if (file_exists(XOOPS_ROOT_PATH . '/modules/' . $moduleDirName . '/include/ratefile.inc.php')) {
        require XOOPS_ROOT_PATH . '/modules/' . $moduleDirName . '/include/ratefile.inc.php';
    }
    trigger_error('not updated rate :');
    exit();
}
//-------------------------------------
//view start
$articles = [];
$category = [];
//module entry data handler
/** @var \XoopsModules\Soapbox\EntrygetHandler $entrydataHandler */
$entrydataHandler = new \XoopsModules\Soapbox\EntrygetHandler();
if (empty($articleID)) {
    //get entry object
    $entryobArray = $entrydataHandler->getArticlesAllPermcheck(1, 0, true, true, 0, 0, null, $sortname, $sortorder, null, null, true, false);
    //    $totalarts = $entrydataHandler->total_getArticlesAllPermcheck;
    if (empty($entryobArray) || 0 === count($entryobArray)) {
        redirect_header(XOOPS_URL . '/modules/' . $moduleDirName . '/index.php', 1, _MD_SOAPBOX_NOTHING);
    }
    $_entryob = $entryobArray[0];
} else {
    //get entry object
    $_entryob = $entrydataHandler->getArticleOnePermcheck($articleID, true, true);
    if (!is_object($_entryob)) {
        redirect_header(XOOPS_URL . '/modules/' . $moduleDirName . '/index.php', 1, 'Not Found');
    }
}
//-------------------------------------
$articles = $_entryob->toArray();
//get category object
$_categoryob = $_entryob->_sbcolumns;
//get vars
$category = $_categoryob->toArray();
//-------------------------------------
//update count
$entrydataHandler->getUpArticlecount($_entryob, true);

//assign
$articles['id']     = $articles['articleID'];
$articles['posted'] = $myts->htmlSpecialChars(formatTimestamp($articles['datesub'], $helper->getConfig('dateformat')));

// includes code by toshimitsu
if ('' !== trim($articles['bodytext'])) {
    $articletext    = explode('[pagebreak]', $_entryob->getVar('bodytext', 'none'));
    $articles_pages = count($articletext);
    if ($articles_pages > 1) {
        require_once XOOPS_ROOT_PATH . '/class/pagenav.php';
        $pagenav = new \XoopsPageNav($articles_pages, 1, $startpage, 'page', 'articleID=' . $articles['articleID']);
        $xoopsTpl->assign('pagenav', $pagenav->renderNav());
        if (0 === $startpage) {
            $articles['bodytext'] = $articles['lead'] . '<br><br>' . $myts->displayTarea($articletext[$startpage], $articles['html'], $articles['smiley'], $articles['xcodes'], 1, $articles['breaks']);
        } else {
            $articles['bodytext'] = &$myts->displayTarea($articletext[$startpage], $articles['html'], $articles['smiley'], $articles['xcodes'], 1, $articles['breaks']);
        }
    } else {
        $articles['bodytext'] = $articles['lead'] . '<br><br>' . $myts->displayTarea($_entryob->getVar('bodytext', 'none'), $articles['html'], $articles['smiley'], $articles['xcodes'], 1, $articles['breaks']);
    }
}
//Cleantags
$articles['bodytext'] = $cleantags->cleanTags($articles['bodytext']);

if (1 === $helper->getConfig('includerating')) {
    $xoopsTpl->assign('showrating', '1');
    //-------------------------------------
    //for ratefile update by domifara
    $xoopsTpl->assign('rate_gtickets', $GLOBALS['xoopsSecurity']->getTokenHTML());
    //-------------------------------------
    if (0.0000 != $articles['rating']) {
        $articles['rating'] = '' . _MD_SOAPBOX_RATING . ': ' . $myts->htmlSpecialChars(number_format($articles['rating'], 2));
        $articles['votes']  = '' . _MD_SOAPBOX_VOTES . ': ' . $myts->htmlSpecialChars($articles['votes']);
    } else {
        $articles['rating'] = _MD_SOAPBOX_NOTRATED;
    }
}

if (is_object($xoopsUser)) {
    $xoopsTpl->assign('authorpm_link', "<a href=\"javascript:openWithSelfMain('" . XOOPS_URL . '/pmlite.php?send2=1&amp;to_userid=' . $category['author'] . "', 'pmlite', 450, 380);\"><img src='" . $pathIcon16 . "/mail_new.png' alt=\"" . _MD_SOAPBOX_WRITEAUTHOR . '"></a>');
} else {
    $xoopsTpl->assign('user_pmlink', '');
}
// Теги
if (xoops_getModuleOption('usetag', 'soapbox')) {
    $moduleHandler = xoops_getHandler('module');
    $tagsModule    = $moduleHandler->getByDirname('tag');
    if (is_object($tagsModule)) {
        require_once XOOPS_ROOT_PATH . '/modules/tag/include/tagbar.php';

        $itemid = \Xmf\Request::getInt('articleID', 0, 'GET');
        $catid  = 0;
        $tagbar = tagBar($itemid, $catid);
        if ($tagbar) {
            $xoopsTpl->assign('tagbar', $tagbar);
            $tagsmeta = implode(' ', $tagbar['tags']);
        } else {
            $tagsmeta = '';
        }
    } else {
        $xoopsTpl->assign('tagbar', false);
        $tagsmeta = '';
    }
}
//if ( xoops_getModuleOption( 'usetag', 'soapbox') ){
//  require_once XOOPS_ROOT_PATH . '/modules/tag/include/tagbar.php';
//  $xoopsTpl->assign( 'tags', true );
//  $xoopsTpl->assign( 'tagbar', tagBar( $_REQUEST['articleID'], 0 ) );
//} else {
//  $xoopsTpl->assign( 'tags', false );
//}

// Functional links
$articles['adminlinks'] = $entrydataHandler->getadminlinks($_entryob, $_categoryob);
$articles['userlinks']  = $entrydataHandler->getuserlinks($_entryob);

$articles['author']     = Soapbox\Utility::getLinkedUnameFromId($category['author'], 0);
$articles['authorname'] = Soapbox\Utility::getAuthorName($category['author']);
$articles['colname']    = $category['name'];
$articles['coldesc']    = $category['description'];
$articles['colimage']   = $category['colimage'];

$xoopsTpl->assign('xoops_pagetitle', $articles['headline']);
$xoopsTpl->assign('story', $articles);
//-----------------------------
$mbmail_subject = sprintf(_MD_SOAPBOX_INTART, $xoopsConfig['sitename']);
$mbmail_body    = sprintf(_MD_SOAPBOX_INTARTFOUND, $xoopsConfig['sitename']);
$al             = Soapbox\Utility::getAcceptLang();
if ('ja' === $al) {
    if (function_exists('mb_convert_encoding') && function_exists('mb_encode_mimeheader')
        && @mb_internal_encoding(_CHARSET)) {
        $mbmail_subject = mb_convert_encoding($mbmail_subject, 'SJIS', _CHARSET);
        $mbmail_body    = mb_convert_encoding($mbmail_body, 'SJIS', _CHARSET);
    }
}
$mbmail_subject = rawurlencode($mbmail_subject);
$mbmail_body    = rawurlencode($mbmail_body);
//-----------------------------
$xoopsTpl->assign('mail_link', 'mailto:?subject=' . $myts->htmlSpecialChars($mbmail_subject) . '&amp;body=' . $myts->htmlSpecialChars($mbmail_body) . ':  ' . XOOPS_URL . '/modules/' . $moduleDirName . '/article.php?articleID=' . $articles['articleID']);
$xoopsTpl->assign('articleID', $articles['articleID']);
$xoopsTpl->assign('lang_ratethis', _MD_SOAPBOX_RATETHIS);
$xoopsTpl->assign('lang_modulename', $xoopsModule->name());
$xoopsTpl->assign('lang_moduledirname', $moduleDirName);
$xoopsTpl->assign('imgdir', $myts->htmlSpecialChars($helper->getConfig('sbimgdir')));
$xoopsTpl->assign('uploaddir', $myts->htmlSpecialChars($helper->getConfig('sbuploaddir')));

//-------------------------------------
//box view
$listarts = [];
//-------------------------------------
$_other_entryob_arr = $entrydataHandler->getArticlesAllPermcheck((int)$helper->getConfig('morearts'), 0, true, true, 0, 0, null, $sortname, $sortorder, $_categoryob, $articles['articleID'], true, false);
$totalartsbyauthor  = (int)$entrydataHandler->total_getArticlesAllPermcheck + 1;

if (!empty($_other_entryob_arr)) {
    foreach ($_other_entryob_arr as $_other_entryob) {
        $link = [];
        $link = $_other_entryob->toArray();
        //--------------------
        $link['id']        = $link['articleID'];
        $link['arttitle']  = $_other_entryob->getVar('headline');
        $link['published'] = $myts->htmlSpecialChars(formatTimestamp($_other_entryob->getVar('datesub'), $helper->getConfig('dateformat')));
        //        $link['poster'] = XoopsUserUtility::getUnameFromId( $link['uid'] );
        $link['poster']      = Soapbox\Utility::getLinkedUnameFromId($category['author']);
        $link['bodytext']    = xoops_substr($link['bodytext'], 0, 255);
        $listarts['links'][] = $link;
    }
    $xoopsTpl->assign('listarts', $listarts);
    $xoopsTpl->assign('readmore', "<a style='font-size: 9px;' href=" . XOOPS_URL . '/modules/' . $moduleDirName . '/column.php?columnID=' . $articles['columnID'] . '>' . _MD_SOAPBOX_READMORE . '[' . $totalartsbyauthor . ']</a> ');
}

if (isset($GLOBALS['xoopsModuleConfig']['globaldisplaycomments'])
    && 1 === $GLOBALS['xoopsModuleConfig']['globaldisplaycomments']) {
    if (1 === $articles['commentable']) {
        require XOOPS_ROOT_PATH . '/include/comment_view.php';
    }
} else {
    require XOOPS_ROOT_PATH . '/include/comment_view.php';
}
$xoopsTpl->assign('xoops_module_header', '<link rel="stylesheet" type="text/css" href="' . XOOPS_URL . '/modules/' . $moduleDirName . '/assets/css/style.css">');

require_once XOOPS_ROOT_PATH . '/footer.php';
